web/src/message_events.js

```diff
@@ -35,6 +37,129 @@ import * as unread_ops from "./unread_ops";
 import * as unread_ui from "./unread_ui";
 import * as util from "./util";
 
+export function update_views_filtered_on_message_property(message_ids, term_type) {
+    // NOTE: Call this function after updating the message property locally.
+    assert(!term_type.includes("not-"));
+
+    // List of narrow terms whose msg list doesn't get updated elsewhere but
+    // can be applied locally.
+    const supported_term_types = [
+        "has-image",
+        "has-link",
+        "has-reaction",
+        "has-attachment",
+        "is-starred",
+        "is-unread",
+        "is-mentioned",
+        "is-alerted",
+        // TODO: Implement support for these terms.
+        // "is-followed",
+    ];
+
+    if (message_ids.length === 0 || !supported_term_types.includes(term_type)) {
+        return;
+    }
+
+    for (const msg_list of message_lists.all_rendered_message_lists()) {
+        const filter = msg_list.data.filter;
+        const filter_term_types = filter.sorted_term_types();
+        if (
+            // Check if current filter relies on the changed message property.
+            !filter_term_types.includes(term_type) &&
+            !filter_term_types.includes(`not-${term_type}`)
+        ) {
+            continue;
+        }
+
+        if (!filter.can_apply_locally()) {
+            // We can't apply the update locally, so we need to rerender the view.
+            const current_selected_id = msg_list.selected_id();
+            message_view.show(filter.terms(), {
+                then_select_id: current_selected_id,
+                trigger: "message property update",
+                force_rerender: true,
+            });
```

Comment:
Should this use a `then_select_offset`? Overall I wonder if we should have an exported method `message_view.force_refresh()` rather than having independent code paths like this one responsible for having the right set of options for that purpose.

Also is this a case that actually occurs? I'm not sure whether all the properties in question have `can_apply_locally`.

Reply:
Oh I guess you could be looking at `is:starred search:foo` and we know the message is now starred, but we don't have a way to check if it matches the search? Though in that case, we could use the API endpoint for asking if something matches a search query rather than doing a full refresh, maybe? That's what we do when new messages arrive while looking at a search view, and it seems potentially applicable here...

Probably worth laying out the reasoning in a comment, even if we don't want to deal with doing that complexity right now.

Reply:
Added a TODO comment for using `/messages/matches_narrow` and combined the `force_rerender` calls so that we are only doing one of of those. I am not sure having a `force_refresh` function make sense since we might again end up calling it with a different set of parameters.